jupyterlab_latex/__init__.py

```diff
@@ -225,6 +228,179 @@ def get(self, path = ''):
                 out = yield self.run_latex(cmd_sequence)
         self.finish(out)
 
+
+class LatexSynctexHandler(APIHandler):
+    """
+    A handler that runs synctex on the server.
+    """
+
+    def initialize(self, notebook_dir):
+        self.notebook_dir = notebook_dir
+
+
+    def build_synctex_edit_cmd(self, pdf_name, pos):
+        """Builds tuple that will be used to call the synctex edit shell command.
+
+        Parameters
+        ----------
+        pdf_name: string
+            This is the name of pdf tex file, without the extension
+
+        pos: dict
+            A dictionary containing the position in the pdf file
+            document to map.
+
+        returns:
+            A dictionary containing the mapped position data.
+
+        """
+        c = LatexConfig(config=self.config)
+        data = f'-i {pos["page"]}:{pos["x"]}:{pos["y"]}:{pdf_name+".pdf"}'
+
+        cmd = (
+            c.synctex_command,
+            'edit',
+            '-o',
+            f'{pos["page"]}:{pos["x"]}:{pos["y"]}:{pdf_name+".pdf"}'
+            )
+
+        return cmd
+
+    def build_synctex_view_cmd(self, tex_name, pos):
+        """Builds tuple that will be used to call the synctex view shell command.
+
+        Parameters
+        ----------
+        tex_name: string
+            This is the base name of the tex file, without the extension
+
+        pos: dict
+            A dictionary containing the position in the tex file
+            document to map.
+
+        returns:
+            A dictionary containing the mapped position data.
+
+        """
+        c = LatexConfig(config=self.config)
+
+        cmd = (
+            c.synctex_command,
+            'view',
+            '-i',
+            f'{pos["line"]}:{pos["column"]}:{tex_name+".tex"}',
+            '-o',
+            f'{tex_name+".pdf"}'
+            )
+
+        return cmd
+
+    def parse_synctex_response(self, response, pos):
+        """
+        Take the stdout response of SyncTex and parse it
+        into a dictionary.
+
+        Parameters
+        ----------
+        response: string
+            The response output to stdout from SyncTeX
+
+        pos: dict
+            The position that was input to SyncTeX
+
+        returns:
+            A dictionary with the parsed response.
+
+        """
+        lines = response.split('\n')
+        started = False
+        synctex_fields = ["line", "column", "page", "x", "y"]
+        result = {}
+        for line in lines:
+            if line == 'SyncTeX result begin':
+                started = True
```

Comment:
Why do we need this started logic? Wouldn't it be better to have a preprocessing step that just takes away everything up to the "SyncTeX result begin" and stops on "SyncTeX result end"?

i think that'd be nice to have as a separate method rather than embedding it here as it makes it substantially more complicated to read.

could call it `isolate_synctex_results(lines)` or something

Reply:
I have no objection to a function that looks like that. Mostly, I didn't want to think up of a regular expression to grab the bits between "SynTeX result begin" and "SynTeX result end" because my regex fu is bad. If you were to suggest one, though...

Reply:
Okay, I have rewritten it to use a regex and a dictionary comprehension instead of a loop. How does it look now?